segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/BaseChunkStorageProvider.java

```diff
@@ -0,0 +1,531 @@
+            Preconditions.checkArgument(!chunks[i].getName().equals(chunks[0].getName()), "source chunk is same as target");
+            Preconditions.checkArgument(!chunks[i].getName().equals(chunks[i - 1].getName()), "duplicate chunk found");
+        }
+    }
+
+    /**
+     * Truncates a given chunk.
+     *
+     * @param handle ChunkHandle of the chunk to truncate.
+     * @param offset Offset to truncate to.
+     * @return True if the object was truncated, false otherwise.
+     * @throws ChunkStorageException Throws ChunkStorageException in case of I/O related exceptions.
+     * @throws UnsupportedOperationException If this operation is not supported by this provider.
+     */
+    @Override
+    final public boolean truncate(ChunkHandle handle, long offset) throws ChunkStorageException, UnsupportedOperationException {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        // Validate parameters
+        Preconditions.checkArgument(null != handle, "handle must not be null");
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be readonly");
+        Preconditions.checkArgument(offset > 0, "handle must not be readonly");
+
+        long traceId = LoggerHelpers.traceEnter(log, "truncate", handle.getChunkName());
+
+        // Call concrete implementation.
+        boolean retValue = doTruncate(handle, offset);
+
+        LoggerHelpers.traceLeave(log, "truncate", traceId, handle.getChunkName());
+
+        return retValue;
+    }
+
+    /**
+     * Sets readonly attribute for the chunk.
+     *
+     * @param handle ChunkHandle of the chunk.
+     * @param isReadonly True if chunk is set to be readonly.
+     * @return True if the operation was successful, false otherwise.
+     * @throws ChunkStorageException Throws ChunkStorageException in case of I/O related exceptions.
+     * @throws UnsupportedOperationException If this operation is not supported by this provider.
+     */
+    @Override
+    final public boolean setReadOnly(ChunkHandle handle, boolean isReadonly) throws ChunkStorageException, UnsupportedOperationException {
+        Exceptions.checkNotClosed(this.closed.get(), this);
+        // Validate parameters
+        Preconditions.checkArgument(null != handle, "handle must not be null");
+
+        long traceId = LoggerHelpers.traceEnter(log, "setReadOnly", handle.getChunkName());
+
+        // Call concrete implementation.
+        boolean retValue = doSetReadOnly(handle, isReadonly);
+
+        LoggerHelpers.traceLeave(log, "setReadOnly", traceId, handle.getChunkName());
+
+        return retValue;
+    }
+
+    /**
+     * Closes.
+     * @throws Exception In case of any error.
+     */
+    @Override
+    public void close() throws Exception {
+        this.closed.set(true);
+    }
+
+    /**
+     * Checks whether this instance is closed or not.
+     * @return True if this instance is closed, false otherwise.
+     */
+    protected boolean isClosed() {
```

Comment:
Initially I had this.closed as protected member so that derived class can access it.
Close is called by container on shutdown.  
So instead of exposing the field I am just exposing it as isClosed() method. 
There is no need to expose this otherwise.

Reply:
What happens if some code tries to use a closed provider?

Reply:
All BaseChunkStorageProvider methods check this to prevent executing code on closed instance.
`Exceptions.checkNotClosed(this.closed.get(), this);`

HDFS provider also uses isClosed but on second thought it now seems redundant .
I'll probably remove it.

```
private void ensureInitializedAndNotClosed() {
        Exceptions.checkNotClosed(this.isClosed(), this);
        Preconditions.checkState(this.fileSystem != null, "HDFSStorage is not initialized.");
    }
```

```
@Override
    public void close() {
        if (!isClosed()) {
            if (this.fileSystem != null) {
                try {
                    this.fileSystem.close();
                    this.fileSystem = null;
                } catch (IOException e) {
                    log.warn("Could not close the HDFS filesystem: {}.", e);
                }
            }
        }
    }
```
@eolivelli Did you find isClosed useful at all ? Otherwise I'll just remove it.

Reply:
@sachin-j-joshi 
I think that this `isClosed()`  method is not useful to the real implementation.
because if the implementation needs to handle its own close protocol it must do it internally and it cannot rely on this flag.

So if this `isClosed()` method is not useful anywhere else I would drop it.

We must **require** from the implementation that you can call `close()` as many times as you want.
We may expect  from the implementation that any calls to a closed implementation instance fails.

btw having our own guard in `BaseCheckStoreProvider `is needed because we have to handle the resources managed by this class, but it is a internal detail, that we should hide from the user.

Reply:
ok.